### PR TITLE
Fix ssl protocol config for ssl context

### PR DESCRIPTION
**Pull request by martinfurmanski**
Using the raw ssl context instead of the netty
handlers missed the configuration of protocols.

The ssl testing is broken out into submodules
and a new exhaustive set of negotiation
circumstances are tested.

### CODE REVIEW

File: integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        @Override
+        protected void channelRead0( ChannelHandlerContext ctx, ByteBuf msg ) throws Exception
+        {
+            collectedData.writeBytes( msg );
+        }
+
+        @Override
+        public void exceptionCaught( ChannelHandlerContext ctx, Throwable cause ) throws Exception
+        {
+            //cause.printStackTrace();
```
Comment (andrewkerr9000) [inline integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java]: Did you mean to leave this commented out?

File: integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        }
+    }
+
+    public static class ClientInitializer extends ChannelInitializer<SocketChannel>
+    {
+        private SslContext sslContext;
+        private final Bucket bucket;
+        private io.netty.util.concurrent.Future<Channel> handshakeFuture;
```
Comment (andrewkerr9000) [inline integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java]: don't think this needs a fully qualified name

File: integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    @Override
+                    protected void initChannel( SocketChannel ch ) throws Exception
+                    {
+                        ChannelPipeline pipeline = ch.pipeline();
+
+                        SSLEngine sslEngine = sslContext.newEngine( ch.alloc() );
+                        sslEngine.setNeedClientAuth( true );
+                        SslHandler sslHandler = new SslHandler( sslEngine );
+                        pipeline.addLast( sslHandler );
+                        // sslHandler.handshakeFuture().addListener( f -> f.cause().printStackTrace() );
```
Comment (andrewkerr9000) [inline integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java]: Commented out line

File: integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        @Override
+        protected void channelRead0( ChannelHandlerContext ctx, ByteBuf msg ) throws Exception
+        {
+            ctx.channel().writeAndFlush( ctx.alloc().buffer().writeBytes( RESPONSE ) );
+        }
+
+        @Override
+        public void exceptionCaught( ChannelHandlerContext ctx, Throwable cause ) throws Exception
+        {
+            //cause.printStackTrace();
```
Comment (andrewkerr9000) [inline integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java]: also commented out line

**Review by andrewkerr9000 — COMMENTED** [review]
Minor comments, mostly commented out lines.

File: integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java (hunk 1 above)
Comment (martinfurmanski) [reply integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java]: Yes, that is just for the future if someone wants to debug using prints. It is meant to function as a hint.

File: integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java (hunk 2 above)
Comment (martinfurmanski) [reply integrationtests/src/test/java/org/neo4j/ssl/SecureClient.java]: Right, was probably needed before I factored the code out.

File: integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java (hunk 3 above)
Comment (martinfurmanski) [reply integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java]: Yea, I tend to leave this just as a hint for an easy way of enabling a print for debugging.

File: integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java (hunk 4 above)
Comment (martinfurmanski) [reply integrationtests/src/test/java/org/neo4j/ssl/SecureServer.java]: Same.

**Comment by martinfurmanski** [comment]
@andrewkerr9000 I've added comments to the commented out debugging prints, and resolved your point around the FQ Future. PTAL!

**Review by andrewkerr9000 — APPROVED** [review]
LGTM

**Comment by zhenlineo** [comment]
👍

**Pull request merged**